[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89188
* #88923
* #88879
* #88668
* __->__ #88665
* #88661

**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @leslie-fang-intel @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
[Quant][FX] Add backend config for onednn backend

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/qconfig_mapping.py:
```diff
@@ -105,6 +105,10 @@ def _get_default_qconfig_mapping(is_qat: bool, backend: str, version: int) -> QC
         .set_object_type(torch.nn.LayerNorm, qconfig_layernorm) \
 
+    if backend == 'onednn':
+        qconfig_mapping.set_object_type(torch.nn.LeakyReLU, qconfig) \
```
do we support quantization for standalone leakyrelu module/op?

Xia-Weiwen replied:
I think it's required that the fused module shares the same `qconfig` with separate modules/ops otherwise there will be an error in UT. So, I added these here.

jerryzh168 replied:
OK, sounds good, we should be able to configure qconfig for patterns I think, but this will come up a bit later, cc @andrewor14 as FYI

jerryzh168 replied:
can you add a TODO comment here? so that we will remember to remove this later when we support fusion patterns in QConfigMapping

Xia-Weiwen replied:
OK, I have added a TODO comment in this PR: https://github.com/pytorch/pytorch/pull/89188

jerryzh168 commented on torch/ao/quantization/backend_config/onednn.py:
```diff
@@ -0,0 +1,189 @@
+    return BackendConfig("onednn") \
+        .set_backend_pattern_configs(conv_configs) \
+        .set_backend_pattern_configs(linear_configs) \
+        .set_backend_pattern_configs(_get_binary_op_configs(binary_op_dtype_configs)) \
+        .set_backend_pattern_config(_get_cat_config(default_op_dtype_configs)) \
+        .set_backend_pattern_configs(_get_default_op_configs(default_op_dtype_configs)) \
+        .set_backend_pattern_configs(_get_fixed_qparams_op_configs(fixed_qparams_op_dtype_configs)) \
+        .set_backend_pattern_configs(_get_share_qparams_op_configs(share_qparams_op_dtype_configs)) \
+        .set_backend_pattern_configs(_get_bn_configs(default_op_dtype_configs)) \
+        .set_backend_pattern_configs(_get_rnn_op_configs(rnn_op_dtype_configs)) \
+        .set_backend_pattern_configs(_get_embedding_op_configs(embedding_op_dtype_configs))
```
are you sure onednn actually supports all the ops here?

Xia-Weiwen replied:
Well, I thought we needed all these configs set so I just copied them here. I will try to remove them.

Xia-Weiwen replied:
They are removed

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"

**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"

**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -363,6 +365,48 @@ def forward(self, x):
             expected_node_occurrence=expected_occurrence)
 
+    @skipIfNoONEDNN
+    def test_fuse_linear_bn_leaky_relu_onednn(self):
```
I have moved tests for fusion from the next PR to this PR

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"

**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168. Do you have more comments? Thanks!

Xia-Weiwen commented:
Hi @jerryzh168. Is it ok to land? Thanks

Xia-Weiwen commented on torch/ao/quantization/backend_config/onednn.py:
```diff
@@ -0,0 +1,152 @@
+    """
+    Return the `BackendConfig` for PyTorch's native ONEDNN backend.
+    """
+    return BackendConfig("onednn") \
+        .set_backend_pattern_configs(conv_configs) \
+        .set_backend_pattern_configs(linear_configs)
```
For other ops, do I need to copy default configs here or just ignore them? https://github.com/pytorch/pytorch/pull/88665#discussion_r1020668935

Xia-Weiwen replied:
Looks like they are needed. I have added them back. Please take a look again. Thanks.

jerryzh168 replied:
OK looks good, so for the other ops do we just fallback to the default (fbgemm) implementation?

Xia-Weiwen replied:
Yes, we use default native implementations for ops other than conv/linear. However, we still need to set pattern configs for other ops here. Otherwise, those ops are not quantized if we use onednn's backend config for `prepare_fx` and `convert_fx`.

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/backend_config/onednn.py:
```diff
@@ -0,0 +1,199 @@
+# 1.1 linear module + leaky_relu fusion config
+# linear leaky_relu, linear module + leaky_relu module
+linear_configs.append(
+    BackendPatternConfig((nn.LeakyReLU, nn.Linear))
```
nit: we updated the pattern format recently, please take a look at https://github.com/pytorch/pytorch/pull/90698

Xia-Weiwen replied:
Thanks. I made changes (using the new format) in this PR: https://github.com/pytorch/pytorch/pull/89188

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant][FX] Add backend config for onednn backend and fuse Linear-LeakyReLU"


**Summary**
Add backend config for onednn backend so that it can support more post op fusion for int8 inference. First `Linear - LeakyReLU` fusion is implemented based on previous PRs.

**Test plan**
python test_quantization.py TestFuseFx

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
@pytorchbot merge
